packages/astro/src/vite-plugin-astro-server/route.ts

```diff
@@ -35,11 +33,9 @@ interface MatchedRoute {
 	mod: ComponentInstance;
 }
 
-function getCustom404Route({ config }: AstroSettings, manifest: ManifestData) {
-	// For Windows compat, use relative page paths to match the 404 route
-	const relPages = resolvePages(config).href.replace(config.root.href, '');
-	const pattern = new RegExp(`${appendForwardSlash(relPages)}404.(astro|md)`);
-	return manifest.routes.find((r) => r.component.match(pattern));
+function getCustom404Route(manifest: ManifestData): RouteData | undefined {
+	const route404 = /^\/404\/?$/;
+	return manifest.routes.find((r) => route404.test(r.route));
```

Comment:
I think this is safe? Previously we only supported `404.astro`, `404.md`, `404.mdx`, and `404.mdoc` as custom 404 routes. Those would all have a `route` of `/404` or `/404/`, so are still supported. This change adds custom 404 support for `src/pages/404.html` or for `injectRoute('404', ...)` in dev.